core/src/SYCL/Kokkos_SYCL_Parallel_Team.hpp

```diff
@@ -556,12 +529,281 @@ class ParallelReduce<FunctorType, Kokkos::TeamPolicy<Properties...>,
+                                 &results_ptr[global_id * value_count]);
+                }
+              }
+              item.barrier(sycl::access::fence_space::local_space);
+
+              // Perform the actual workgroup reduction. To achieve a better
+              // memory access pattern, we use sequential addressing and a
+              // reversed loop. If the workgroup size is 8, the first element
+              // contains all the values with index%4==0, after the second one
+              // the values with index%2==0 and after the third one index%1==0,
+              // i.e., all values.
+              for (unsigned int stride = wgroup_size / 2; stride > 0;
+                   stride >>= 1) {
+                const auto idx = local_id;
+                if (idx < stride) {
+                  ValueJoin::join(selected_reducer,
+                                  &local_mem[idx * value_count],
+                                  &local_mem[(idx + stride) * value_count]);
+                }
+                item.barrier(sycl::access::fence_space::local_space);
+              }
+
+              // Finally, we copy the workgroup results back to global memory to
+              // be used in the next iteration. If this is the last iteration,
+              // i.e., there is only one workgroup also call final() if
+              // necessary.
+              if (local_id == 0) {
+                ValueOps::copy(
+                    functor,
+                    &results_ptr2[(item.get_group_linear_id()) * value_count],
+                    &local_mem[0]);
+                if constexpr (ReduceFunctorHasFinal<FunctorType>::value)
+                  if (n_wgroups <= 1 && item.get_group_linear_id() == 0) {
+                    FunctorFinal<FunctorType, WorkTag>::final(
+                        static_cast<const FunctorType&>(functor),
+                        &results_ptr2[(item.get_group_linear_id()) *
+                                      value_count]);
+                  }
+              }
+            });
+      });
+      space.fence();
+
+      // FIXME_SYCL this is likely not necessary, see above
+      Kokkos::Impl::DeepCopy<Kokkos::Experimental::SYCLDeviceUSMSpace,
+                             Kokkos::Experimental::SYCLDeviceUSMSpace>(
+          space, results_ptr, results_ptr2,
+          sizeof(*m_result_ptr) * value_count * n_wgroups);
+      space.fence();
+
+      first_run = false;
+      size      = n_wgroups;
+    }
+
+    // At this point, the reduced value is written to the entry in results_ptr
+    // and all that is left is to copy it back to the given result pointer if
+    // necessary.
+    if (m_result_ptr) {
+      Kokkos::Impl::DeepCopy<Kokkos::Experimental::SYCLDeviceUSMSpace,
+                             Kokkos::Experimental::SYCLDeviceUSMSpace>(
+          space, m_result_ptr, results_ptr,
+          sizeof(*m_result_ptr) * value_count);
+      space.fence();
+    }
+
+    sycl::free(results_ptr, q);
```

Comment:
You forgot to free `results_ptr2`.  Put these in a `std::unique_ptr` with a custom deleter.

Reply:
Bummer. I thought I caught all of them in the various versions. :slightly_frowning_face: